cpp/oneapi/dal/detail/communicator.cpp

```diff
@@ -0,0 +1,172 @@
+/*******************************************************************************
+* Copyright 2020-2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/detail/communicator.hpp"
+
+namespace oneapi::dal::detail::v1 {
+
+const char* communication_error::what() const noexcept {
+    return std::runtime_error::what();
+}
+
+spmd_request::spmd_request() : impl_(nullptr) {}
+
+void spmd_request::wait() {
+    if (impl_) {
+        return impl_->wait();
+    }
+}
+
+bool spmd_request::test() {
+    if (impl_) {
+        return impl_->test();
+    }
+    return true;
+}
+
+std::int64_t spmd_communicator::get_rank() const {
+    return impl_->get_rank();
+}
+
+std::int64_t spmd_communicator::get_rank_count() const {
+    return impl_->get_rank_count();
+}
+
+std::int64_t spmd_communicator::get_default_root_rank() const {
+    return impl_->get_default_root_rank();
+}
+
+void spmd_communicator::barrier() const {
+    return impl_->barrier();
+}
+
+spmd_request spmd_communicator::bcast(byte_t* send_buf,
```

Comment:
why don't use simple void* instead of byte_t*?

Reply:
`void*` is discouraged in modern C++. The interface assumes memory buffer, so it accepts array of bytes as the most convenient low-level representation. 

BTW, you will not be required to cast typed buffers to `byte_t*`, there are template versions of these function in hpp.

Reply:
ok